### PR TITLE
Add `FactoredDiphoneBlockV2` with right context output for training

### PR DESCRIPTION
**Pull request by NeoLegends**
This PR adds an output on `p(r | c, x)`. This usually gives some small WER improvements if you also have a loss on this in training.

When decoding w/ the joint output, the right context is ignored (as we are still using a diphone model).

V2 to avoid hash breakage.

### CODE REVIEW

**Review by Marvin84 — APPROVED** [review]

File: i6_models/parts/factored_hybrid/diphone.py [hunk 1]
```diff
@@ -148,3 +150,64 @@ def forward_joint(self, features: Tensor) -> Tensor:
...
+    def forward_factored(
+        self,
+        features: Tensor,  # B, T, F
+        contexts_left: Tensor,  # B, T
+        contexts_center: Tensor,  # B, T
+    ) -> Tuple[Tensor, Tensor, Tensor, Tensor, Tensor]:
+        """
+        :param features: Main encoder output. shape B, T, F. F=num_inputs
+        :param contexts_left: The left contexts used to compute p(c|l,x), shape B, T.
+        :param contexts_center: The center states used to compute p(r|c,x), shape B, T.
```
Comment (michelwi) [inline i6_models/parts/factored_hybrid/diphone.py:202]: It is sort of implicit in your naming scheme (e.g. of `center_state_embedding` vs `left_context_embedding`) But there is iirc the difference that for `l` and `r` we only predict/encode the phoneme identity but `c` consists of the center phoneme and the state index. This is I think automatically handled via `get_center_dim`.

Should/can this be better documented in the docstrings? 
E.g. I do not find it obvious that `contexts_left` can be from `[1, num_contexts]` while `contexts_center` should be from `[1, some_factor * num_contexts]`. Could this be a source of (user) errors or is it otherwise also obvious in our implementations?

File: i6_models/parts/factored_hybrid/diphone.py [hunk 2]
```diff
@@ -148,3 +150,64 @@ def forward_joint(self, features: Tensor) -> Tensor:
...
+        :return: tuple of logits for p(c|l,x), p(l|x), p(r|c,x) and the embedded left context and center state values.
+        """
+
+        logits_center, logits_left, contexts_embedded_left = super().forward_factored(features, contexts_left)
+
+        # in training we forward exactly one context per T, so: B, T, E
+        center_states_embedded = self.center_state_embedding(contexts_center)
+        features_right = torch.cat((features, center_states_embedded), -1)  # B, T, F+E
+        logits_right = self.right_context_encoder(features_right)
```
[the remark this answers is not in the document]
Comment (NeoLegends) [reply i6_models/parts/factored_hybrid/diphone.py:211]: Correct.

Even the decoding can remain the same, as you can (successfully) run decodings w/ only the diphone part of a triphone model.

If you wanted I could also add that variant (i.e. the triphone) to this PR.

[the remark this answers is not in the document]
Comment (michelwi) [reply i6_models/parts/factored_hybrid/diphone.py:211]: > This usually gives some small WER improvements if you also have a loss on this in training.

Did you ever try how "factored diphone with auxilliary right context loss" compares to "factored triphone but in recognition we only use diphone"?

File: i6_models/parts/factored_hybrid/diphone.py (hunk 1 above)
Comment (NeoLegends) [reply i6_models/parts/factored_hybrid/diphone.py]: Good point. I'll extend the documentation.

File: i6_models/parts/factored_hybrid/diphone.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (NeoLegends) [reply i6_models/parts/factored_hybrid/diphone.py:219]: Yes. IIRC the difference was non-existent or minimal, but I'd have to look up the exact results.
